### PR TITLE
Refactor Workload 

### PR DESCRIPTION
**Pull request by samliok**
I approached this refactor by considering how a VM developer would want to test their VM. This PR is not the final refactor of the testing framework for the HyperSDK, but I believe it's a good starting point to discuss what that framework _should_ be.

## Changes

- Reduces the interface VM developers need to implement to just one method that generates transactions. This is a significant simplification compared to the previous `TxWorkloadFactory`, which required implementing two functions that, in turn, required the user to implement `TxWorkloadIterators` with an additional two functions.

```go
type TxAssertion func(ctx context.Context, require *require.Assertions, uri string)
type TxGenerator interface {
    GenerateTx(context.Context, string) (*chain.Transaction, TxAssertion, error)
}
```

- Generating "workloads" is now handled under the hood by the `TxWorkload` struct that has one `TxGenerator` field.
- This simplification means all a developer needs to do is implement the `TxGenerator` interface and pass it to the e2e tests to sanity check their VM with the HyperSDK's e2e suite(well that and include a run script, cmd, and other extra stuff that should be abstracted soon).

## Reasoning & Design

1. Current e2e testing served as both HyperSDK tests and tests for MorpheusVM, which added layers of complexity and cognitive load for devs.
2. We already have e2e tests that are testing both the HyperSDK and MorpheusVM. Let's make these e2e tests generic and easily digestible for developers. 
3. We want to offer a variety of testing experiences to developers on the HyperSDK. Procedural testing(#1359) for VMs should be included in addition to this form of testing.

### CODE REVIEW

File: examples/vmwithcontracts/go.mod [hunk 1]
```diff
@@ -7,22 +7,16 @@ require (
...
 	github.com/near/borsh-go v0.3.1
-	github.com/onsi/ginkgo/v2 v2.13.1
 	github.com/spf13/cobra v1.7.0
 	github.com/status-im/keycard-go v0.2.0
 	github.com/stretchr/testify v1.8.4
 )
 
 require (
 	filippo.io/edwards25519 v1.0.0 // indirect
+	github.com/BurntSushi/toml v1.3.2 // indirect
```
Comment (richardpringle) [inline examples/vmwithcontracts/go.mod:17]: 😮 @samliok, do you know who this is?

Comment (samliok) [reply examples/vmwithcontracts/go.mod]: seems `go mod tidy` removed this

File: examples/morpheusvm/tests/generator/mixed.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+type multiAuthTxGenerator struct {
+	txCheckInterval     time.Duration
+	addressAndFactories []addressAndFactory
+	balance             uint64
+	count               int
+}
+
+func NewMultiAuthTxGenerator(key *fixture.Ed25519TestKey, txCheckInterval time.Duration) (workload.TxGenerator, error) {
+	blsPriv, err := bls.GeneratePrivateKey()
```
Comment (samliok) [inline examples/morpheusvm/tests/generator/mixed.go:35]: i don't know how these bls keys were previously able to send balances. They are randomly generated and I can't seem to find where we allocate their balances.

File: examples/morpheusvm/cmd/morpheus-cli/cmd/chain.go [hunk 3]
```diff
@@ -37,7 +37,7 @@ var chainInfoCmd = &cobra.Command{
 
 var watchChainCmd = &cobra.Command{
 	Use: "watch",
-	RunE: func(_ *cobra.Command, args []string) error {
+	RunE: func(_ *cobra.Command, _ []string) error {
```
Comment (tsachiherman) [inline examples/morpheusvm/cmd/morpheus-cli/cmd/chain.go:40]: that's 100% ok, but not really related to this PR.

**Review by tsachiherman — CHANGES_REQUESTED** [review]
Maybe I'm missing some file, but I can't find any use of `NewMultiAuthTxGenerator` or `multiAuthTxGenerator`.

File: tests/workload/transactions.go [hunk 4]
```diff
@@ -106,23 +61,41 @@ func GenerateNBlocks(ctx context.Context, require *require.Assertions, uris []st
...
 }
 
-func GenerateUntilStop(
+// GenerateTxs generates transactions using the provided TxGenerator until the generator
+// can no longer generate transactions
+// ClientUri is the uri of the client that will generate the transactions
+// ConfirmUris is a list of uris to confirm the transactions
+func (w *TxWorkload) GenerateTxs(ctx context.Context, require *require.Assertions, amount int, clientURI string, confirmUris []string) {
+	// TODO: why do we only use the first uri for submitting transactions when it differs from the confirmUris?
+	submitClient := jsonrpc.NewJSONRPCClient(confirmUris[0])
```
[the remark this answers is not in the document]
Comment (samliok) [reply tests/workload/transactions.go:70]: should we load balance this selection?

[the remark this answers is not in the document]
Comment (aaronbuchwald) [reply tests/workload/transactions.go:70]: I don't think that's necessary

**Review by aaronbuchwald — DISMISSED** [review]
LGTM - we can add back the functionality of `Workloads(...)` via https://github.com/ava-labs/hypersdk/pull/1667

File: tests/workload/transactions.go [hunk 5]
```diff
@@ -15,79 +15,34 @@ import (
...
 
 	_, startHeight, _, err := client.Accepted(ctx)
 	require.NoError(err)
 	height := startHeight
-	targetheight := startHeight + n
+	targetheight := startHeight + uint64(blocks)
 
-	for generator.Next() && height < targetheight {
-		tx, confirm, err := generator.GenerateTxWithAssertion(ctx)
+	for count := 0; count < blocks && height < targetheight; count++ {
```
[the remark this answers is not in the document]
Comment (samliok) [reply tests/workload/transactions.go:40]: I could extend this function to take in another argument `txPerBlock`, this is the same as the previous logic for generating blocks. Is it worth doing this is a follow up?

[the remark this answers is not in the document]
Comment (aaronbuchwald) [reply tests/workload/transactions.go:40]: This was actually a looser condition before, not more exact by requiring a specific number of transactions. We can produce blocks with no transactions in them and this function is intended to guarantee at least that number of blocks were produced, but only checks the height to allow empty blocks to be included.

Could we remove the extra check and leave this as only `height < targetHeight` ?

**Review by aaronbuchwald — APPROVED** [review]

**Review by tsachiherman — APPROVED** [review]

**Pull request merged**